File: nn/src/test/scala/com/intel/webscaleml/nn/nn/performTest/ClassNLLCriterionPerform.lua
--
-- Created by yao on 6/3/16
--

--require 'torch'
require 'nn'

forwardIterations = 10
backwardIterations = 10
inputNum = 512
featureDim = 512
seed = 100
torch.manualSeed(seed)
file = io.open("TorchPerform.csv", "a")
io.output(file)

output = torch.Tensor(inputNum, featureDim)
target = torch.Tensor(inputNum)
target:apply(function()
    i = torch.ceil(torch.uniform(1, 10))
    return i
end)

criterion = nn.ClassNLLCriterion()

--warm up
for j = 1, forwardIterations do
    criterion:forward(output, target)
end

timer = torch.Timer()
for j = 1, forwardIterations do
    criterion:forward(output, target)
end
timeSpend = timer:time().real / forwardIterations
file:write("ClassNLLCriterion_forward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")

timer = torch.Timer()
for j = 1, backwardIterations do
    criterion:backward(output, target)
end
timeSpend = timer:time().real / backwardIterations
file:write("ClassNLLCriterion_backward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")

file:close()

File: nn/src/test/scala/com/intel/webscaleml/nn/nn/performTest/ClassNLLCriterionPerform.scala
package com.intel.webscaleml.nn.nn.performTest

import com.intel.webscaleml.nn.nn.ClassNLLCriterion
import com.intel.webscaleml.nn.tensor.{RandomGenerator, torch}
import org.scalatest.{BeforeAndAfter, FlatSpec}

/**
  * Created by yao on 5/19/16.
  */
class ClassNLLCriterionPerform extends FlatSpec with BeforeAndAfter{
  val ForwardIterations = PerformTestData.forwardIterations
  val BackwardIterations = PerformTestData.backwardIterations
  val ForwardTimeoutMillis = 1.0
  val BackwardTimeoutMillis = 1.0
  val InputNum = 512
  val FeatureDim = 512
  val Seed = 100
  val Generator = new RandomGenerator(Seed)
  val isRun = TestUtils.isRun()

  val criterion = new ClassNLLCriterion[Float]()
  val input = torch.Tensor[Float](InputNum, FeatureDim).rand()
  val target = torch.Tensor[Float](InputNum)

  //Initialize output and targets
  target.apply1(_ => Generator.uniform(1, 10).asInstanceOf[Float].ceil)

  //Warm up JVM
  for (j <- 0 until ForwardIterations) {
    criterion.forward(input, target)
  }

  "ClassNLLCriterion " should "forward with a good performance" in {
    if (!TestUtils.isRun())
      cancel("Performance regression tests are forbidden")
    val timeMillis = TestUtils.testCriterionForwardPerform(input, target, criterion, ForwardIterations)
    assert(timeMillis < ForwardTimeoutMillis)
  }

  it should " backward with a good performance" in {
    if (!TestUtils.isRun())
      cancel("Performance regression tests are forbidden")
    val timeMillis = TestUtils.testCriterionBackwardPerform(input, target, criterion, BackwardIterations)
    assert(timeMillis < BackwardTimeoutMillis)
  }

}
